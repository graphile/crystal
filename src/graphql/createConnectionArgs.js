import { memoize, fromPairs, includes, snakeCase, toUpper } from 'lodash'
import { GraphQLEnumType, GraphQLInt, GraphQLBoolean } from 'graphql'
import { CursorType } from './types.js'
import getType from './getType.js'

const createConnectionArgs = (table, ignoreColumnConditions = []) => ({
  // The column specified by `orderBy` means more than just the order to
  // return items in. This column is also the column we will use for
  // cursors.
  orderBy: {
    type: createTableOrderingEnum(table),
    description:
      'The order the resulting items should be returned in. This argument ' +
      'is also important as it is used in creating pagination cursors. This ' +
      'value’s default is the primary key for the object.',
    defaultValue: (() => {
      const column = table.getPrimaryKeys()[0]
      if (column) return column.name
      return null
    })(),
  },
  first: {
    type: GraphQLInt,
    description:
      'The top `n` items in the set to be returned. Can’t be used ' +
      'with `last`.',
  },
  last: {
    type: GraphQLInt,
    description:
      'The bottom `n` items in the set to be returned. Can’t be used ' +
      'with `first`.',
  },
  before: {
    type: CursorType,
    description: 'Constrains the set to nodes *before* this cursor in the specified ordering.',
  },
  after: {
    type: CursorType,
    description: 'Constrains the set to nodes *after* this cursor in the specified ordering.',
  },
  offset: {
    type: GraphQLInt,
    description: 'An integer offset representing how many items to skip in the set.',
  },
  descending: {
    type: GraphQLBoolean,
    description:
      'If `true` the nodes will be in descending order, if `false` the ' +
      'items will be in ascending order. `false` by default.',
    defaultValue: false,
  },
  ...fromPairs(
    table
    .getColumns()
    // If `ignoreColumnConditions` is set to true, all column conditions will
    // be disabled. If `ignoreColumnConditions` is an array, only certain
    // conditions will be ignored.
    .filter(column => (ignoreColumnConditions === true ? false : !includes(ignoreColumnConditions, column)))
    .map(column => [column.getFieldName(), {
      type: getType(column.type),
      description:
        'Filters the resulting set with an equality test on the ' +
        `${column.getMarkdownFieldName()} field.`,
      // TODO: Deprecate this…
      // deprecationReason:
      //   'Simple equality testing is insufficient for the nodes field and just ' +
      //   'adds noise. Instead use procedures for custom set filtering.',
    }])
  ),
})

export default createConnectionArgs

/**
 * Creates an ordering enum which simply contains all of a `Table`s columns.
 *
 * @param {Table} table
 * @returns {GraphQLEnumType}
 */
// TODO: Some way to eliminate some columns from ordering enum?
export const createTableOrderingEnum = memoize(table =>
  new GraphQLEnumType({
    name: `${table.getTypeName()}Ordering`,
    description: `Properties with which ${table.getMarkdownTypeName()} can be ordered.`,

    values: fromPairs(
      table
      .getColumns()
      .map(column => [toUpper(snakeCase(column.getFieldName())), {
        value: column.name,
        description: column.description,
      }])
    ),
  }))
