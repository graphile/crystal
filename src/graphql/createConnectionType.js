import { memoize } from 'lodash'
import createTableType from './createTableType.js'
import { PageInfoType, CursorType } from './types.js'

import {
  GraphQLObjectType,
  GraphQLList,
  GraphQLNonNull,
  GraphQLInt,
} from 'graphql'

const createTableConnectionType = memoize(table =>
  new GraphQLObjectType({
    name: `${table.getTypeName()}Connection`,
    description: `A connection to a list of ${table.getMarkdownTypeName()} items`,

    // TODO: Implement a `ConnectionType` interface

    fields: {
      pageInfo: {
        type: new GraphQLNonNull(PageInfoType),
        description: `Information to aid in pagination of type ${table.getMarkdownTypeName()}.`,
        resolve: pageInfo => pageInfo,
      },
      totalCount: {
        type: GraphQLInt,
        description: 'All of the items available to be queried in this connection.',
        resolve: ({ totalCount }) => totalCount,
      },
      nodes: {
        type: new GraphQLList(createTableType(table)),
        description: `The queried list of ${table.getMarkdownTypeName()}.`,
        resolve: ({ nodes }) => nodes,
      },
      edges: {
        type: new GraphQLList(createTableEdgeType(table)),
        description: 'A single item and a cursor to aid in pagination.',
        resolve: ({ edges }) => edges,
      },
    },
  })
)

export default createTableConnectionType

const createTableEdgeType = table =>
  new GraphQLObjectType({
    name: `${table.getTypeName()}Edge`,
    description: `An edge in the \`${table.getTypeName()}Connection\`.`,

    fields: {
      cursor: {
        type: new GraphQLNonNull(CursorType),
        description: 'The cursor describing the position of the edgeâ€™s associated node.',
        resolve: ({ cursor }) => cursor || 'null',
      },
      node: {
        type: createTableType(table),
        description: 'The item at the end of the edge.',
        resolve: ({ node }) => node,
      },
    },
  })
