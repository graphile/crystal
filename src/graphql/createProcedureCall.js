const createProcedureCall = procedure => {
  // Construct the qualified procedure name.
  const procedureName = `"${procedure.schema.name}"."${procedure.name}"`

  // Construct the argument list for the procedure call.
  const procedureArgsList = Array.from(procedure.args).map(([name, type], i) => {
    const placeholder = `$${i + 1}`

    // If the type of this argument is a table type, we will be expecting JSON
    // but we need that JSON to be a table. Therefore we run it through
    // `json_populate_record`.
    if (type.isTableType)
      return `json_populate_record(null::"${type.table.schema.name}"."${type.table.name}", ${placeholder})`

    return placeholder
  }).join(', ')

  // Add the procedure name with the procedure argument list.
  const procedureCall = `${procedureName}(${procedureArgsList})`

  return procedureCall
}

export default createProcedureCall
