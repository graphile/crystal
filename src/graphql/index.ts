import createGraphQLSchema from './schema/createGQLSchema'

export {
  createGraphQLSchema,
}
