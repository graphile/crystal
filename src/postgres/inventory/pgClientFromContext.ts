// TODO: Refactor this module, it has code smellâ€¦
import { mixed } from '../../interfaces';
import { ClientBase, PoolClient } from 'pg';

export const $$pgClient = 'pgClient';

/**
 * Retrieves a Postgres client from a context, throwing an error if such a
 * client does not exist.
 */
export default function getPgClientFromContext(context: mixed): PoolClient {
  if (context == null || typeof context !== 'object') throw new Error('Context must be an object.');

  const client = context[$$pgClient];

  if (client == null) throw new Error('Postgres client does not exist on the context.');

  if (!(client instanceof ClientBase))
    throw new Error('Postgres client on context is of the incorrect type.');

  return client as PoolClient;
}
