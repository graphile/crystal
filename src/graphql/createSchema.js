import { GraphQLSchema } from 'graphql'
import createQueryType from './query/createQueryType.js'
import createMutationType from './mutation/createMutationType.js'

/**
 * Creates a GraphQLSchema from a PostgreSQL schema.
 *
 * @param {Schema} schema
 * @returns {GrpahQLSchema}
 */
const createGraphqlSchema = (schema, options) =>
  new GraphQLSchema({
    query: createQueryType(schema, options),
    mutation: createMutationType(schema, options),
  })

export default createGraphqlSchema
