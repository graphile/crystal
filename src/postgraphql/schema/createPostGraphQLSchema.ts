import { Pool, Client, ClientConfig, connect as connectPgClient } from 'pg'
import { GraphQLSchema, GraphQLOutputType } from 'graphql'
import { Inventory, Type, ObjectType, getNonNullableType } from '../../interface'
import { introspectPgDatabase, addPgCatalogToInventory } from '../../postgres'
import { PgCatalog, PgCatalogClass, PgCatalogProcedure } from '../../postgres/introspection'
import getTypeFromPgType from '../../postgres/inventory/type/getTypeFromPgType'
import PgClassType from '../../postgres/inventory/type/PgClassType'
import { createGraphQLSchema } from '../../graphql'
import BuildToken from '../../graphql/schema/BuildToken'
import createPgProcedureMutationGqlFieldEntry from './procedures/createPgProcedureMutationGqlFieldEntry'
import createPgProcedureQueryGqlFieldEntry from './procedures/createPgProcedureQueryGqlFieldEntry'
import createPgProcedureObjectTypeGqlFieldEntry from './procedures/createPgProcedureObjectTypeGqlFieldEntry'
import getPgProcedureComputedClass from './procedures/getPgProcedureComputedClass'
import getPgTokenTypeFromIdentifier from './auth/getPgTokenTypeFromIdentifier'
import getJwtGqlType from './auth/getJwtGqlType'

/**
 * Creates a PostGraphQL schema by looking at a Postgres client.
 */
export default async function createPostGraphQLSchema (
  clientOrConfig?: Pool | Client | ClientConfig | string,
  schemaOrCatalog: string | Array<string> | PgCatalog = 'public',
  options: {
    classicIds?: boolean,
    dynamicJson?: boolean,
    jwtSecret?: string,
    jwtPgTypeIdentifier?: string,
    disableDefaultMutations?: boolean,
  } = {},
): Promise<GraphQLSchema> {
  // Create our inventory.
  const inventory = new Inventory()
  let pgCatalog: PgCatalog

  // If schema already is a PgCatalog use it, otherwise 'build' the PgCatalog
  if (schemaOrCatalog instanceof PgCatalog) {
    pgCatalog = schemaOrCatalog
  }
  else {
    // If our argument was not an array, make it one.
    const schemas = Array.isArray(schemaOrCatalog) ? schemaOrCatalog : [schemaOrCatalog]

    // Introspect our Postgres database to get a catalog. If we weren’t given a
    // client, we will just connect a default client from the `pg` module.
    // TODO: test
    if (clientOrConfig instanceof Client || clientOrConfig instanceof Pool) {
      const pgClient = clientOrConfig
      pgCatalog = await introspectPgDatabase(pgClient, schemas)
    }
    else {
      const pgClient = await connectPgClient(clientOrConfig || {})
      pgCatalog = await introspectPgDatabase(pgClient, schemas)
      pgClient.end()
    }
  }

  // Gets the Postgres token type from our provided identifier. Just null if
  // we don’t have a token type identifier.
  const jwtPgType = options.jwtPgTypeIdentifier
    ? getPgTokenTypeFromIdentifier(pgCatalog, options.jwtPgTypeIdentifier)
    : undefined

  // If a token type is defined, but the JWT secret is not. Throw an error.
  if (jwtPgType && !options.jwtSecret)
    throw new Error('Postgres token type is defined, but a JWT secret is not defined. Please provide a JWT secret.')

  // Add all of our Postgres constructs to that inventory.
  addPgCatalogToInventory(inventory, pgCatalog, {
    renameIdToRowId: options.classicIds,
  })

  // TODO: Move all procedure and procedure hook code somewhere else…

  // Create “sinks,” or places that our Postgres procedures will go. Each
  // “sink” represents a different location in our GraphQL schema where the
  // procedure may be exposed.
  const pgMutationProcedures: Array<PgCatalogProcedure> = []
  const pgQueryProcedures: Array<PgCatalogProcedure> = []
  const pgObjectTypeProcedures: Map<PgCatalogClass, Array<PgCatalogProcedure>> = new Map()

  // For all of the procedures in our catalog, find a place to put each one.
  for (const pgProcedure of pgCatalog.getProcedures()) {
    // If this procedure is unstable, it is a mutation. Add it to that list.
    if (!pgProcedure.isStable)
      pgMutationProcedures.push(pgProcedure)
    else {
      // Detect if this procedure is a computed procedure.
      const pgComputedClass = getPgProcedureComputedClass(pgCatalog, pgProcedure)

      // If it is not a computed procedure, add it to the normal query
      // procedure list.
      if (!pgComputedClass)
        pgQueryProcedures.push(pgProcedure)
      // If this does have a class for which it is a computed procedure, add it
      // to that class’s procedure array.
      else {
        // If this class does not yet have an array of procedures, create one.
        if (!pgObjectTypeProcedures.has(pgComputedClass))
          pgObjectTypeProcedures.set(pgComputedClass, [])

        // Actually add the procedure.
        pgObjectTypeProcedures.get(pgComputedClass)!.push(pgProcedure)
      }
    }
  }

  // Actually create our GraphQL schema.
  const gqlSchema = createGraphQLSchema(inventory, {
    nodeIdFieldName: options.classicIds ? 'id' : 'nodeId',
    dynamicJson: options.dynamicJson,
    disableDefaultMutations: options.disableDefaultMutations,

    // If we have a JWT Postgres type, let us override the GraphQL output type
    // with our own.
    _typeOverrides: jwtPgType && new Map<Type<mixed>, { input?: true, output?: GraphQLOutputType }>([
      [getTypeFromPgType(pgCatalog, jwtPgType), {
        // Throw an error if the user tries to use this as input.
        get input (): never { throw new Error(`Using the JWT Token type '${options.jwtPgTypeIdentifier}' as input is not yet implemented.`) },
        // Use our JWT GraphQL type as the output.
        output: getJwtGqlType(getNonNullableType(getTypeFromPgType(pgCatalog, jwtPgType)) as PgClassType, options.jwtSecret!),
      }],
    ]),

    _hooks: {
      // Extra field entries to go on the mutation type.
      mutationFieldEntries: _buildToken =>
        pgMutationProcedures.map(pgProcedure => createPgProcedureMutationGqlFieldEntry(_buildToken, pgCatalog, pgProcedure)),

      // Extra field entries to go on the query type.
      queryFieldEntries: _buildToken =>
        pgQueryProcedures.map(pgProcedure => createPgProcedureQueryGqlFieldEntry(_buildToken, pgCatalog, pgProcedure)),

      // Extra field entires to go on object types that also happen to be
      // classes.
      objectTypeFieldEntries: <TValue>(objectType: ObjectType<TValue>, _buildToken: BuildToken) =>
        objectType instanceof PgClassType
          ? (pgObjectTypeProcedures.get(objectType.pgClass) || []).map(pgProcedure => createPgProcedureObjectTypeGqlFieldEntry(_buildToken, pgCatalog, pgProcedure))
          : [],
    },
  })

  return gqlSchema
}
