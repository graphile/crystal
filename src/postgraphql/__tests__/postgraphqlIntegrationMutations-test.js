import { resolve as resolvePath } from 'path'
import { readFile, readdirSync } from 'fs'
import { graphql } from 'graphql'
import withPgClient from '../../__tests__/utils/withPgClient'
import { $$pgClient } from '../../postgres/inventory/pgClientFromContext'
import createPostGraphQLSchema from '../schema/createPostGraphQLSchema'

// This test suite can be flaky. Increase itâ€™s timeout.
jasmine.DEFAULT_TIMEOUT_INTERVAL = 1000 * 20

const kitchenSinkData = new Promise((resolve, reject) => {
  readFile('examples/kitchen-sink/data.sql', (error, data) => {
    if (error) reject(error)
    else resolve(data.toString().replace(/begin;|commit;/g, ''))
  })
})

const mutationsDir = resolvePath(__dirname, 'fixtures/mutations')
const mutationFileNames = readdirSync(mutationsDir)
let mutationResults = []

beforeAll(() => {
  // Get a GraphQL schema instance that we can query.
  const gqlSchemaPromise = withPgClient(async pgClient => {
    return await createPostGraphQLSchema(pgClient, ['a', 'b', 'c'])
  })()

  // Execute all of the mutations in parallel. We will not wait for them to
  // resolve or reject. The tests will do that.
  //
  // All of our mutations get there own Postgres client instance. Queries share
  // a client instance.
  mutationResults = mutationFileNames.map(async fileName => {
    // Wait for the schema to resolve. We need the schema to be introspected
    // before we can do anything else!
    let gqlSchema = await gqlSchemaPromise
    // Get a new Postgres client and run the mutation.
    return await withPgClient(async pgClient => {
      // Read the mutation from the file system.
      const mutation = await new Promise((resolve, reject) => {
        readFile(resolvePath(mutationsDir, fileName), 'utf8', (error, data) => {
          if (error) reject(error)
          else resolve(data)
        })
      })

      // Add data to the client instance we are using.
      await pgClient.query(await kitchenSinkData)

      // Return the result of our GraphQL query.
      return await graphql(gqlSchema, mutation, null, { [$$pgClient]: pgClient })
    })()
  })
})

for (let i = 0; i < mutationFileNames.length; i++) {
  test(mutationFileNames[i], async () => {
    expect(await mutationResults[i]).toMatchSnapshot()
  })
}
