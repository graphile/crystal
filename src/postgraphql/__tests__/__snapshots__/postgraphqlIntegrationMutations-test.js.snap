exports[`test mutation-create.graphql 1`] = `
Object {
  "data": Object {
    "a": Object {
      "clientMutationId": null,
      "query": Object {
        "nodeId": "query",
      },
      "type": Object {
        "anIntRange": Object {
          "end": Object {
            "inclusive": false,
            "value": 500,
          },
          "start": null,
        },
        "bigint": "467131188225",
        "boolean": false,
        "compoundType": Object {
          "a": 123,
          "b": "abc",
          "c": "GREEN",
          "d": "ec4a9fae-4ec5-4763-98eb-0327fb2dc9bf",
          "fooBar": 20,
        },
        "date": "2016-10-15",
        "daterange": Object {
          "end": Object {
            "inclusive": false,
            "value": "1927-11-07",
          },
          "start": Object {
            "inclusive": true,
            "value": "1927-11-06",
          },
        },
        "decimal": 15.2,
        "domain": 6,
        "domain2": 5,
        "enum": "RED",
        "id": 201,
        "interval": Object {
          "days": 4,
          "hours": 3,
          "minutes": 2,
          "months": 5,
          "seconds": 1,
          "years": 6,
        },
        "json": "{\"x\":1,\"y\":2,\"z\":3}",
        "jsonb": "{\"a\":1,\"b\":2,\"c\":3}",
        "money": 5000,
        "nestedCompoundType": Object {
          "a": Object {
            "a": 456,
            "b": "def",
            "c": "BLUE",
            "d": "79863dcf-0433-4c3d-bc51-978326d4546f",
            "fooBar": 42,
          },
          "b": Object {
            "a": 789,
            "b": "ghi",
            "c": "RED",
            "d": "b687ee42-c515-4544-b742-525e39517e7d",
            "fooBar": -8,
          },
          "bazBuz": 0,
        },
        "nodeId": "WyJ0eXBlcyIsMjAxXQ==",
        "numeric": 15.2,
        "numrange": Object {
          "end": null,
          "start": Object {
            "inclusive": true,
            "value": 50,
          },
        },
        "smallint": 30,
        "textArray": Array [
          "have",
          "you",
          "ever",
          "been",
          "down",
          "the",
          "water",
          "spout",
          "?",
        ],
        "time": "19:13:18.625699",
        "timestamp": "2016-10-07T16:12:21.747269",
        "timestamptz": "2016-10-09T16:12:45.218676-04:00",
        "timetz": "13:13:29.585176-04",
        "varchar": "abc",
      },
    },
    "b": Object {
      "a": Object {
        "cursor": null,
        "node": Object {
          "name": "John Smith Jr.",
          "nodeId": "WyJwZW9wbGUiLDkwMDBd",
        },
      },
      "b": Object {
        "cursor": null,
        "node": Object {
          "name": "John Smith Jr.",
          "nodeId": "WyJwZW9wbGUiLDkwMDBd",
        },
      },
      "c": Object {
        "cursor": null,
        "node": Object {
          "name": "John Smith Jr.",
          "nodeId": "WyJwZW9wbGUiLDkwMDBd",
        },
      },
      "clientMutationId": null,
      "d": Object {
        "cursor": null,
        "node": Object {
          "name": "John Smith Jr.",
          "nodeId": "WyJwZW9wbGUiLDkwMDBd",
        },
      },
      "e": Object {
        "cursor": null,
        "node": Object {
          "name": "John Smith Jr.",
          "nodeId": "WyJwZW9wbGUiLDkwMDBd",
        },
      },
      "f": Object {
        "cursor": null,
        "node": Object {
          "name": "John Smith Jr.",
          "nodeId": "WyJwZW9wbGUiLDkwMDBd",
        },
      },
      "g": Object {
        "cursor": null,
        "node": Object {
          "name": "John Smith Jr.",
          "nodeId": "WyJwZW9wbGUiLDkwMDBd",
        },
      },
      "person": Object {
        "about": "Son of Sara and John Smith.",
        "email": "johnny.boy.smith@email.com",
        "id": 9000,
        "name": "John Smith Jr.",
        "nodeId": "WyJwZW9wbGUiLDkwMDBd",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "c": Object {
      "a": Object {
        "cursor": null,
        "node": Object {
          "name": "Best Pal",
          "nodeId": "WyJwZW9wbGUiLDIwXQ==",
        },
      },
      "b": Object {
        "cursor": null,
        "node": Object {
          "name": "Best Pal",
          "nodeId": "WyJwZW9wbGUiLDIwXQ==",
        },
      },
      "c": Object {
        "cursor": null,
        "node": Object {
          "name": "Best Pal",
          "nodeId": "WyJwZW9wbGUiLDIwXQ==",
        },
      },
      "clientMutationId": "hello",
      "d": Object {
        "cursor": null,
        "node": Object {
          "name": "Best Pal",
          "nodeId": "WyJwZW9wbGUiLDIwXQ==",
        },
      },
      "e": Object {
        "cursor": null,
        "node": Object {
          "name": "Best Pal",
          "nodeId": "WyJwZW9wbGUiLDIwXQ==",
        },
      },
      "f": Object {
        "cursor": null,
        "node": Object {
          "name": "Best Pal",
          "nodeId": "WyJwZW9wbGUiLDIwXQ==",
        },
      },
      "g": Object {
        "cursor": null,
        "node": Object {
          "name": "Best Pal",
          "nodeId": "WyJwZW9wbGUiLDIwXQ==",
        },
      },
      "person": Object {
        "about": "My archnemisis is Budd Deey.",
        "email": "best.pal@email.com",
        "id": 20,
        "name": "Best Pal",
        "nodeId": "WyJwZW9wbGUiLDIwXQ==",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "d": Object {
      "clientMutationId": "world",
      "compoundKey": Object {
        "extra": false,
        "nodeId": "WyJjb21wb3VuZF9rZXlzIiw5MDAwLDIwXQ==",
        "personByPersonId1": Object {
          "name": "John Smith Jr.",
          "nodeId": "WyJwZW9wbGUiLDkwMDBd",
        },
        "personByPersonId2": Object {
          "name": "Best Pal",
          "nodeId": "WyJwZW9wbGUiLDIwXQ==",
        },
        "personId1": 9000,
        "personId2": 20,
      },
      "personByPersonId1": Object {
        "name": "John Smith Jr.",
        "nodeId": "WyJwZW9wbGUiLDkwMDBd",
      },
      "personByPersonId2": Object {
        "name": "Best Pal",
        "nodeId": "WyJwZW9wbGUiLDIwXQ==",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "e": Object {
      "edgeCase": Object {
        "notNullHasDefault": true,
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "f": Object {
      "edgeCase": Object {
        "notNullHasDefault": false,
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "g": Object {
      "a": Object {
        "cursor": null,
        "node": Object {
          "name": "Budd Deey",
          "nodeId": "WyJwZW9wbGUiLDE5OThd",
        },
      },
      "b": Object {
        "cursor": null,
        "node": Object {
          "name": "Budd Deey",
          "nodeId": "WyJwZW9wbGUiLDE5OThd",
        },
      },
      "c": Object {
        "cursor": null,
        "node": Object {
          "name": "Budd Deey",
          "nodeId": "WyJwZW9wbGUiLDE5OThd",
        },
      },
      "clientMutationId": null,
      "d": Object {
        "cursor": null,
        "node": Object {
          "name": "Budd Deey",
          "nodeId": "WyJwZW9wbGUiLDE5OThd",
        },
      },
      "e": Object {
        "cursor": null,
        "node": Object {
          "name": "Budd Deey",
          "nodeId": "WyJwZW9wbGUiLDE5OThd",
        },
      },
      "f": Object {
        "cursor": null,
        "node": Object {
          "name": "Budd Deey",
          "nodeId": "WyJwZW9wbGUiLDE5OThd",
        },
      },
      "g": Object {
        "cursor": null,
        "node": Object {
          "name": "Budd Deey",
          "nodeId": "WyJwZW9wbGUiLDE5OThd",
        },
      },
      "person": Object {
        "about": null,
        "email": "budd.deey.the.second@email.com",
        "id": 1998,
        "name": "Budd Deey",
        "nodeId": "WyJwZW9wbGUiLDE5OThd",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
  },
}
`;

exports[`test mutation-delete.graphql 1`] = `
Object {
  "data": Object {
    "a": Object {
      "clientMutationId": null,
      "deletedPostId": "WyJwb3N0cyIsMV0=",
      "post": Object {
        "authorId": 2,
        "headline": "No… It’s a thing; it’s like a plan, but with more greatness.",
        "id": 1,
        "nodeId": "WyJwb3N0cyIsMV0=",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "b": Object {
      "clientMutationId": "hello",
      "deletedPostId": "WyJwb3N0cyIsMl0=",
      "post": Object {
        "authorId": 1,
        "headline": "I hate yogurt. It’s just stuff with bits in.",
        "id": 2,
        "nodeId": "WyJwb3N0cyIsMl0=",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "c": null,
    "d": Object {
      "clientMutationId": "world",
      "deletedPostId": "WyJwb3N0cyIsM10=",
      "post": Object {
        "authorId": 1,
        "headline": "Is that a cooking show?",
        "id": 3,
        "nodeId": "WyJwb3N0cyIsM10=",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "e": Object {
      "clientMutationId": null,
      "deletedPostId": "WyJwb3N0cyIsNl0=",
      "post": Object {
        "authorId": 3,
        "headline": "Stop talking, brain thinking. Hush.",
        "id": 6,
        "nodeId": "WyJwb3N0cyIsNl0=",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "f": Object {
      "clientMutationId": "hello",
      "deletedPostId": "WyJwb3N0cyIsOV0=",
      "post": Object {
        "authorId": 3,
        "headline": "You know how I sometimes have really brilliant ideas?",
        "id": 9,
        "nodeId": "WyJwb3N0cyIsOV0=",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "g": null,
    "h": Object {
      "clientMutationId": "world",
      "deletedPostId": "WyJwb3N0cyIsMTFd",
      "post": Object {
        "authorId": 3,
        "headline": "They’re not aliens, they’re Earth…liens!",
        "id": 11,
        "nodeId": "WyJwb3N0cyIsMTFd",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "i": Object {
      "clientMutationId": null,
      "compoundKey": Object {
        "nodeId": "WyJjb21wb3VuZF9rZXlzIiw0LDNd",
        "personByPersonId1": Object {
          "name": "Kathryn Ramirez",
          "nodeId": "WyJwZW9wbGUiLDRd",
        },
        "personByPersonId2": Object {
          "name": "Budd Deey",
          "nodeId": "WyJwZW9wbGUiLDNd",
        },
        "personId1": 4,
        "personId2": 3,
      },
      "deletedCompoundKeyId": "WyJjb21wb3VuZF9rZXlzIiw0LDNd",
      "query": Object {
        "nodeId": "query",
      },
    },
    "j": Object {
      "clientMutationId": null,
      "compoundKey": Object {
        "nodeId": "WyJjb21wb3VuZF9rZXlzIiwyLDNd",
        "personByPersonId1": Object {
          "name": "Sara Smith",
          "nodeId": "WyJwZW9wbGUiLDJd",
        },
        "personByPersonId2": Object {
          "name": "Budd Deey",
          "nodeId": "WyJwZW9wbGUiLDNd",
        },
        "personId1": 2,
        "personId2": 3,
      },
      "deletedCompoundKeyId": "WyJjb21wb3VuZF9rZXlzIiwyLDNd",
      "query": Object {
        "nodeId": "query",
      },
    },
    "k": Object {
      "clientMutationId": null,
      "deletedPersonId": "WyJwZW9wbGUiLDNd",
      "person": Object {
        "email": "budd.deey@email.com",
        "id": 3,
        "name": "Budd Deey",
        "nodeId": "WyJwZW9wbGUiLDNd",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
  },
  "errors": Array [
    [GraphQLError: No values were deleted in collection 'posts' because no values were found.],
    [GraphQLError: No values were deleted in collection 'posts' because no values were found.],
  ],
}
`;

exports[`test mutation-update.graphql 1`] = `
Object {
  "data": Object {
    "a": Object {
      "clientMutationId": null,
      "person": Object {
        "about": "An older John Smith",
        "email": "john.smith@email.com",
        "id": 1,
        "name": "John Smith Sr.",
        "nodeId": "WyJwZW9wbGUiLDFd",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "b": Object {
      "clientMutationId": "hello",
      "person": Object {
        "about": null,
        "email": "sarah.smith@email.com",
        "id": 2,
        "name": "Sarah Smith",
        "nodeId": "WyJwZW9wbGUiLDJd",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "c": Object {
      "clientMutationId": "world",
      "person": Object {
        "about": "Now with an “H.”",
        "email": "sarah.smith@email.com",
        "id": 2,
        "name": "Sarah Smith",
        "nodeId": "WyJwZW9wbGUiLDJd",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "d": Object {
      "clientMutationId": null,
      "person": Object {
        "about": "I have taken over Budd’s account. Hehehe.",
        "email": "budd.deey@email.com",
        "id": 3,
        "name": "Best Pal",
        "nodeId": "WyJwZW9wbGUiLDNd",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "e": Object {
      "clientMutationId": null,
      "person": Object {
        "about": "Make art not friends.",
        "email": "kathryn.ramirez@email.com",
        "id": 4,
        "name": "Kathryn Ramirez",
        "nodeId": "WyJwZW9wbGUiLDRd",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "f": Object {
      "clientMutationId": null,
      "compoundKey": Object {
        "extra": true,
        "nodeId": "WyJjb21wb3VuZF9rZXlzIiwyLDJd",
        "personByPersonId1": Object {
          "id": 2,
          "name": "Sarah Smith",
        },
        "personByPersonId2": Object {
          "id": 2,
          "name": "Sarah Smith",
        },
        "personId1": 2,
        "personId2": 2,
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "g": Object {
      "clientMutationId": "hello",
      "compoundKey": Object {
        "extra": false,
        "nodeId": "WyJjb21wb3VuZF9rZXlzIiwzLDJd",
        "personByPersonId1": Object {
          "id": 3,
          "name": "Best Pal",
        },
        "personByPersonId2": Object {
          "id": 2,
          "name": "Sarah Smith",
        },
        "personId1": 3,
        "personId2": 2,
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "h": Object {
      "clientMutationId": "world",
      "compoundKey": Object {
        "extra": false,
        "nodeId": "WyJjb21wb3VuZF9rZXlzIiw0LDNd",
        "personByPersonId1": Object {
          "id": 4,
          "name": "Kathryn Ramirez",
        },
        "personByPersonId2": Object {
          "id": 3,
          "name": "Best Pal",
        },
        "personId1": 4,
        "personId2": 3,
      },
      "query": Object {
        "nodeId": "query",
      },
    },
    "i": Object {
      "clientMutationId": null,
      "person": Object {
        "about": null,
        "email": "sarah.smith@email.com",
        "id": 2,
        "name": "Sarah Smith",
        "nodeId": "WyJwZW9wbGUiLDJd",
      },
      "query": Object {
        "nodeId": "query",
      },
    },
  },
}
`;

exports[`test procedure-mutation.graphql 1`] = `
Object {
  "data": Object {
    "add1Mutation": Object {
      "clientMutationId": null,
      "integer": 3,
    },
    "add2Mutation": Object {
      "clientMutationId": "hello",
      "integer": 4,
    },
    "add3Mutation": Object {
      "clientMutationId": "world",
      "integer": null,
    },
    "add4Mutation": Object {
      "integer": 4,
    },
    "compoundTypeMutation": Object {
      "compoundType": Object {
        "a": 420,
        "b": "easy cheesy baked potatoes",
        "c": "RED",
        "d": null,
        "fooBar": 8,
      },
    },
    "intSetMutation": Object {
      "integers": Array [
        1,
        2,
        3,
        4,
        5,
        null,
        6,
      ],
    },
    "jsonIdentityMutation": Object {
      "json": "{\"a\":1,\"b\":2,\"c\":3}",
    },
    "mult1": Object {
      "integer": 0,
    },
    "mult2": Object {
      "integer": 1,
    },
    "mult3": Object {
      "integer": 2,
    },
    "mult4": Object {
      "integer": 10,
    },
    "noArgsMutation": Object {
      "clientMutationId": "x",
      "integer": 2,
    },
    "returnVoidMutation": Object {
      "__typename": "ReturnVoidMutationPayload",
    },
    "tableMutation": Object {
      "personByAuthorId": Object {
        "id": 5,
        "name": "Joe Tucker",
      },
      "post": Object {
        "authorId": 5,
        "headline": "Please, Don-Bot… look into your hard drive, and open your mercy file!",
        "id": 5,
        "nodeId": "WyJwb3N0cyIsNV0=",
      },
      "postEdge": Object {
        "cursor": null,
        "node": Object {
          "headline": "Please, Don-Bot… look into your hard drive, and open your mercy file!",
          "id": 5,
        },
      },
    },
    "tableSetMutation": Object {
      "people": Array [
        Object {
          "name": "John Smith",
        },
        Object {
          "name": "Sara Smith",
        },
        Object {
          "name": "Budd Deey",
        },
        Object {
          "name": "Kathryn Ramirez",
        },
        Object {
          "name": "Joe Tucker",
        },
      ],
    },
    "typesMutation": Object {
      "boolean": false,
    },
  },
}
`;
